server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierRecordingService.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.cluster.service;
+
+import org.elasticsearch.cluster.service.ClusterApplierRecordingService.Stats.Recording;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.metrics.MeanMetric;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.core.Tuple;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+public final class ClusterApplierRecordingService {
+
+    private final Map<String, MeanMetric> recordedActions = new ConcurrentHashMap<>();
+
+    synchronized Stats getStats() {
+        return new Stats(
+            recordedActions.entrySet().stream()
+                .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, e -> new Recording(e.getValue().count(), e.getValue().sum())))
+        );
+    }
+
+    synchronized void updateStats(Recorder recorder) {
+        Set<String> seenActions = new HashSet<>();
+        for (Tuple<String, Long> entry : recorder.recordings) {
+            String action = entry.v1();
+            long timeSpentMS = entry.v2();
+
+            MeanMetric metric = recordedActions.computeIfAbsent(action, key -> new MeanMetric());
+            metric.inc(timeSpentMS);
+            seenActions.add(action);
+        }
+        recordedActions.entrySet().removeIf(entry -> seenActions.contains(entry.getKey()) == false);
+    }
+
+    static final class Recorder {
+
+        private String currentAction;
+        private long startTimeNS;
+        private boolean recording;
+        private final List<Tuple<String, Long>> recordings = new LinkedList<>();
+
+        Releasable record(String action) {
+            if (recording) {
+                throw new IllegalStateException("already recording");
+            }
+
+            this.recording = true;
+            this.currentAction = action;
+            this.startTimeNS = System.nanoTime();
```

Comment:
I think we're ok with milliseconds so we can use `ThreadPool#rawRelativeTimeInMillis` which will let us simulate the passage of time in tests.